### PR TITLE
Simplify propagators

### PR DESCRIPTION
**Pull request by seemk**
## Changes

* Remove code duplication (make hex string related stuff use methods in `detail` namespace)
* Clean up headers
* Remove use of cout (should not pollute stdout, some programs send their data over it).
* Add missing b3 propagator to Bazel tests.

I'd also like to rename the propagator files for consistency:
* `propagation/b3_propagator.h` -> `propagation/b3.h`
* `propagation/http_trace_context` -> `propagation/w3c.h`
* `propagation/jaeger.h` will stay as is

What do you think?

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [x] Changes in public API reviewed

### CODE REVIEW

File: api/include/opentelemetry/trace/trace_flags.h [hunk 1]
```diff
@@ -31,13 +31,16 @@ class TraceFlags final
...
   static constexpr uint8_t kIsSampled = 1;
+  static constexpr uint8_t kForbidden = 0xFF;
 
   TraceFlags() noexcept : rep_{0} {}
 
   explicit TraceFlags(uint8_t flags) noexcept : rep_(flags) {}
 
   bool IsSampled() const noexcept { return rep_ & kIsSampled; }
 
+  bool IsValid() const noexcept { return rep_ != kForbidden; }
```
Comment (pyohannes) [inline api/include/opentelemetry/trace/trace_flags.h:42]: Where is it specified that `0xff` is an invalid value for trace state?

Comment (seemk) [reply api/include/opentelemetry/trace/trace_flags.h:42]: Ah my bad, thanks! 0xFF was an invalid marker for the version not trace flags. Will fix in the morning, not at the computer atm

File: api/include/opentelemetry/trace/propagation/http_trace_context.h [hunk 2]
```diff
@@ -83,168 +65,79 @@ class HttpTraceContext : public TextMapPropagator<T>
...
-    nostd::string_view trace_flags = trace_parent.substr(
-        kHeaderElementLengths[0] + kHeaderElementLengths[1] + kHeaderElementLengths[2] + 3);
 
-    if (version == "ff" || trace_id == "00000000000000000000000000000000" ||
-        span_id == "0000000000000000")
-    {
-      return SpanContext(false, false);
-    }
+    TraceId trace_id       = TraceIdFromHex(trace_id_hex);
+    SpanId span_id         = SpanIdFromHex(span_id_hex);
```
Comment (pyohannes) [inline api/include/opentelemetry/trace/propagation/http_trace_context.h:132]: I didn't test it, but it seems to me this wouldn't catch too short trace and span ids. Like when parsing a header `00-1-1-00`.

File: api/include/opentelemetry/trace/propagation/detail/hex.h [hunk 3]
```diff
@@ -43,13 +43,13 @@ bool IsValidHex(nostd::string_view s)
  */
 bool HexToBinary(nostd::string_view hex, uint8_t *buffer, size_t buffer_size)
 {
+  std::memset(buffer, 0, buffer_size);
```
Comment (ThomsonTan) [inline api/include/opentelemetry/trace/propagation/detail/hex.h:46]: Also check the input `hex` is empty, as it cannot be handled in below code?

File: api/include/opentelemetry/trace/propagation/http_trace_context.h (hunk 2 above)
Comment (seemk) [reply api/include/opentelemetry/trace/propagation/http_trace_context.h]: Added more checks and additional test cases for w3c propagation.

File: api/include/opentelemetry/trace/trace_flags.h (hunk 1 above)
Comment (seemk) [reply api/include/opentelemetry/trace/trace_flags.h]: Fixed - it now checks the version for 0xFF not trace flags :facepalm:

File: api/include/opentelemetry/trace/propagation/detail/hex.h (hunk 3 above)
Comment (seemk) [reply api/include/opentelemetry/trace/propagation/detail/hex.h:46]: It is handled - nothing is written to the buffer if hex is empty.

**Review by lalitb — APPROVED** [review]
LGTM. Thanks for refactoring this.

**Comment by lalitb** [comment]
quick validated these changes against w3c tracecontext tests (https://github.com/open-telemetry/opentelemetry-cpp/tree/main/ext/test/w3c_tracecontext_test), and don't see any new failure ( other than existing trace-state failures). Will go ahead and merge them now.

**Review by ThomsonTan — APPROVED** [review]

**Pull request merged**